src/redux/request.js

```diff
@@ -0,0 +1,271 @@
+ * In order for request epics to handle actions for a specific
+ * request reducer, the names of each reducer/epic combo must match
+ * exactly. Additonally, it must be unique to a reducer/epic combo,
+ * otherwise other request actions with the same name will trigger
+ * the epic.
+ *
+ * The `serialize` function parameters will be the `data` and
+ * `options` arguments provided to the request action. This should
+ * return serialized data to send with the request (params or body).
+ *
+ * The `normalize` function will simply recieve the payload from the
+ * request's response along with the `options` argument provided to
+ * the initial request's action.
+ *
+ * When `defaultParams` is `false`, no query params will ever be
+ * appended to the URL. To append query params, simply provide an
+ * empty object. Key-values will be used as defaults for the query.
+ *
+ * @param {string} name - name of this request epic; must match a
+ * corrisponding request reducer
+ * @param {string|function} [endpoint=''] - string or generator which when
+ * combined with `okapi.url` will become the request URL
+ * @param {object|false} [defaultParams=false] - when `false` this request has
+ * no query options; when an object is provided, it will be merged with
+ * `data` before being stringified and appended to GET and DELETE URLs
+ * @param {function} [normalize=identity] - transforms data being
+ * received from the response before it is dispatched to the reducer
+ * @param {function} [serialize=identity] - transforms data being sent
+ * with the request before the request is made
+ * @returns {function} epic function to use with redux-observable
+ */
+export function createRequestEpic({
+  name,
+  endpoint = '',
+  defaultParams = false,
+  normalize = (i) => i,
+  serialize = (i) => i
+}) {
+  return (action$, { getState }) => (
+    action$.ofType(REQUEST_MAKE)
+      .filter((action) => action.name === name)
+      .switchMap(({ data = {}, options = {} }) => {
+        let { okapi } = getState();
+        let { method = 'GET' } = options;
+        let body, headers = { 'X-Okapi-Tenant': okapi.tenant };
+        let url = endpoint;
+
+        // endpoint url
+        if (typeof endpoint === 'function') {
+          url = `${okapi.url}/${endpoint(data, options)}`;
+        } else {
+          url = `${okapi.url}/${endpoint}`;
+        }
+
+        // request body
+        if (method === 'PUT' || method === 'POST') {
+          headers['Content-Type'] = 'application/json';
+          body = JSON.stringify(serialize(data));
+
+        // query params
+        } else if (defaultParams) {
+          let query = { ...defaultParams, ...data };
+
+          if (Object.keys(query)) {
+            let searchQuery = '';
+
+            // "hack" to prevent query-string from removing/encoding null searches
+            if (query.search === '%00') {
+              let { search, ...q } = query;
+              searchQuery = `search=${search}&${queryString.stringify(q)}`;
```

Comment:
I know that the RMAPI uses the `search` parameter, but the standard web convention is `q`. It's an issue for our API, but I think we should use it in the URL, and also in our API. (Also, pretty sure that this is what CQL uses)

Reply:
I like `q` much better, but as of right now, there is no mapping between front-end query params and back-end query params (I can add it if it's worth it). 

I think we could just keep this until we implement `q` in our own API. That way we can remove this whole `null` hack and have mod-kb-ebsco handle that instead too.